docs/proposals/runtime-client-server.md

```diff
@@ -0,0 +1,221 @@
+
+## Abstract
+
+A proposal of client/server implementation of kubelet container runtime interface.
+
+## Motivation
+
+Currently, any container runtime has to be linked into the kubelet. This makes
+experimentation difficult, and prevents users from landing an alternate
+container runtime without landing code in core kubernetes.
+
+To facilitate experimentation and to enable user choice, this proposal adds a
+client/server implementation of the [new container runtime interface](https://github.com/kubernetes/kubernetes/pull/25899). The main goal
+of this proposal is:
+
+- make it easy to integrate new container runtimes
+- improve code maintainability
+
+## Proposed design
+
+**Design of client/server container runtime**
+
+The main idea of client/server container runtime is to keep main control logic in kubelet while letting remote runtime only do dedicated actions. An alpha [container runtime API](../../pkg/kubelet/api/v1alpha1/runtime/api.proto) is introduced for integrating new container runtimes. The API is based on [protobuf](https://developers.google.com/protocol-buffers/) and [gPRC](http://www.grpc.io) for a number of benefits:
+
+- Perform faster than json
+- Get client bindings for free: gRPC supports ten languages
+- No encoding/decoding codes needed
+- Manage api interfaces easily: server and client interfaces are generated automatically
+
+A new container runtime manager `KubeletGenericRuntimeManager` will be introduced to kubelet, which will
+
+- conforms to kubelet's [Runtime](../../pkg/kubelet/container/runtime.go#L58) interface
+- manage Pods and Containers lifecycle according to kubelet policies
+- call remote runtime's API to perform specific pod, container or image operations
+
+A simple workflow of invoking remote runtime API on starting a Pod can be shown:
+
+```
+Kubelet                  KubeletGenericRuntimeManager       RemoteRuntime
+   +                              +                               +
+   |                              |                               |
+   +---------SyncPod------------->+                               |
+   |                              |                               |
+   |                              +---- Create PodSandbox ------->+
+   |                              |                               |
+   |                              XXXXXXXXXXXX                    |
+   |                              |          X                    |
+   |                              |    NetworkPlugin.             |
+   |                              |       SetupPod                |
+   |                              |          X                    |
+   |                              XXXXXXXXXXXX                    |
+   |                              |                               |
+   |                              +<------------------------------+
+   |                              +----    Pull image    -------->+
+   |                              +<------------------------------+
```

Comment:
Yes, it is in another commit. check
https://github.com/kubernetes/kubernetes/pull/17048/files#diff-822f0e081c10d8b83d7c2ad1391d55f7R69




2016-07-08 15:11 GMT+08:00 Tim Hockin <notifications@github.com>:

> In docs/proposals/runtime-client-server.md
> <https://github.com/kubernetes/kubernetes/pull/17048#discussion_r70033915>
> :
>
> > +   +                              +                               +
> > +   |                              |                               |
> > +   +---------SyncPod------------->+                               |
> > +   |                              |                               |
> > +   |                              +---- Create PodSandbox ------->+
> > +   |                              |                               |
> > +   |                              XXXXXXXXXXXX                    |
> > +   |                              |          X                    |
> > +   |                              |    NetworkPlugin.             |
> > +   |                              |       SetupPod                |
> > +   |                              |          X                    |
> > +   |                              XXXXXXXXXXXX                    |
> > +   |                              |                               |
> > +   |                              +<------------------------------+
> > +   |                              +----    Pull image    -------->+
> > +   |                              +<------------------------------+
>
> I don't see that - did you push?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17048/files/8219903aa0560ac0c25504963e295c227fb49967#r70033915>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AApTHckv5pswnYGaidkhK544Wnt811-jks5qTfgVgaJpZM4GfQ_B>
> .
>